tests/lib/openstack_query/query_blocks/test_query_builder.py

```diff
@@ -257,39 +204,139 @@ def test_add_filter_no_server_side_filter(instance):
         ),
     ],
 )
-def test_add_filter_with_server_side_filter(
-    set_server_side_filters, server_side_filters, expected_values, instance
+def test_parse_where_when_filter_exists(
+    existing_filters, new_filters, expected_out, instance, run_parse_where_test
 ):
     """
-    Tests add_filter method works properly - with one set of server side filters
-    Should only set client_side_filter
+    Tests parse_where, for various cases where adding new preset requires a new
+    server-side filter to be added, to an already populated server_side_filters property from previous presets
+    Tests that the two filters are concatenated properly
     """
-    mock_client_side_filter = MagicMock()
-    instance.server_side_filters = set_server_side_filters
+    instance.server_side_filters = existing_filters
 
-    # pylint:disable=protected-access
-    instance._add_filter(
-        client_side_filter=mock_client_side_filter,
-        server_side_filters=server_side_filters,
-    )
+    filter1_fallback_func = NonCallableMock()
+    instance.server_filter_fallback = [filter1_fallback_func]
 
-    assert instance.server_side_filters == expected_values
-    assert instance.server_filter_fallback == [mock_client_side_filter]
+    mock_client_filter = NonCallableMock()
+    run_parse_where_test(
+        mock_get_filter_return=new_filters,
+        mock_get_filter_func_return=mock_client_filter,
+        expected_server_side_filters=expected_out,
+        expected_client_side_filters=[],
+        expected_fallback_filters=[filter1_fallback_func, mock_client_filter],
+    )
 
 
-def test_add_filter_conflicting_presets(instance):
+@pytest.mark.parametrize(
+    "existing_filters, new_filters",
+    [
+        # one server-side filter set
+        # same filter to add again - conflicting
+        (
+            [{"filter1": "val1"}],
+            [{"filter1": "val2"}],
+        ),
+        # multiple server-side filters set
```

Comment:
These should be split into test cases and the common assertion / setup code moved into a method

That way if it fails, it comes up with the test name i.e.
`test_add_filter_conflicting_resets_multiple_server_side_filters failed`

Whereas at the moment it will just say this failed and it's not obvious which case would be broken

Reply:
:sob:  noo! But DRY!!
but fair enough - I will do this